### PR TITLE
Allow missing libraries

### PR DESCRIPTION
**Pull request by maxim-belkin**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [ ] Have you successfully run `brew tests` with your changes locally?

-----

This enables formula creators/editors to say that the formula can have broken linkage by adding `allow_missing_libs` to the formula definition.
Things to consider:
1. Shall this be a bit more precise and affect `broken_dylibs` only?
2. Shall this be always `false` on a Mac and define `allow_missing_libs` on Linux only (under `extend/os/linux/linkage_checker`)?

CC @sjackman @iMichka

### CODE REVIEW

**Comment by maxim-belkin** [comment]
Testing this on `root` formula (Homebrew on Linux Docker container):
```sh
$ brew linkage --test root
Missing libraries:
  libASImage.so
  libCling.so
  libCore.so
  libEG.so
  libFTGL.so
  libFoam.so
  libGLEW.so
  libGX11.so
  libGed.so
  libGeom.so
  libGpad.so
  libGraf.so
  libGraf3d.so
  libGui.so
  libHist.so
  libHistFactory.so
  libImt.so
  libMLP.so
  libMathCore.so
  libMathMore.so
  libMatrix.so
  libMinuit.so
  libMultiProc.so
  libNet.so
  libNew.so
  libPhysics.so
  libProof.so
  libProofPlayer.so
  libRCsg.so
  libRGL.so
  libRHTTP.so
  libRIO.so
  libROOTBrowserv7.so
  libROOTDataFrame.so
  libROOTGpadv7.so
  libROOTHist.so
  libROOTHistDraw.so
  libROOTNTuple.so
  libROOTVecOps.so
  libROOTWebDisplay.so
  libRint.so
  libRooFit.so
  libRooFitCore.so
  libRooStats.so
  libTMVA.so
  libThread.so
  libTree.so
  libTreePlayer.so
  libTreeViewer.so
  libWebGui6.so
  libXMLIO.so
  libXMLParser.so
  libXrdProofd.so
  libvdt.so
Broken library linkage ignored

[linuxbrew@docker] [Sun Jun 28 04:45:46] [Last Exit Code: 0]
```

**Comment by iMichka** [comment]
Can we have a mechanism for macOS/Linux? I do not know any of these cases on macOS, and a few on Linux, so this is mosty a Linux-only "issue". The idea would be to have two different lists/functions for both OSes.

**Comment by maxim-belkin** [comment]
I'm not sure I understand the question/request. The way I implemented it here is that by adding `allow_missing_libs` line to formulae that have "broken" linkage should let them pass `brew linkage --test` test. I looked at the output of missing libs for `root` package and thought that listing specific library names would be an ovekill. In the current implementation, formulae in both homebrew-core and linuxbrew-core can do that. I can move things around to make it Linux-only -- I just need to know if this is what is wanted/needed.

**Comment by MikeMcQuaid** [comment]
Can you provide an example usage of what the `root` formula would look like with this?

> 1. Shall this be a bit more precise and affect `broken_dylibs` only?

Yes, I think so. I think this should specify a list of libraries (or even a glob) and there needs to be a match otherwise it's an error (so we are forced to remove these when they no longer reply).

> 2. Shall this be always `false` on a Mac and define `allow_missing_libs` on Linux only (under `extend/os/linux/linkage_checker`)?

If we have no usage on macOS so far: yes, I think so. In that case the DSL should be designed in such a way that we could add it on macOS later if need be and be clear that it only applies to Linux now.

Nice work @maxim-belkin!

**Comment by MikeMcQuaid** [comment]
> (not sure why the above link doesn't render into a code snippet 😕)

Not the same repo.

> In the current implementation, it'll look like this:
> 
> https://github.com/maxim-belkin/linuxbrew-core/blob/823efb5eb087a64c108aa5e5087cf3dcbe3e288c/Formula/root.rb#L57

Gotcha. Yeh, I think this should specify the libraries and have something in the name or a parameter that mentions `linux`.

> Shall it be implemented as a block or similarly to `skip_clean`?

Similar to `skip_clean` works for me. Could even allow providing a regex (which errors if there's 0️⃣ matches).



> OK, I'll update the PR accordingly. But just to make sure we're on the same page: do you expect formulae on macOS to respond `false` to `.allow_missing_libs?` (or whatever other name you may suggest -- I spent 1.3 seconds thinking about it) or fail on that call?

Thanks! I think it depends on whether the name is Linux-y or whether it takes a `linux` parameter. @Homebrew/linux thoughts?

**Comment by iMichka** [comment]
I think we should go with linux in the method's name. We have done the same for on_macos/on_linux, where we have specific methods by OS, and not one generic method with a parameter.

By the way, I think the `root` libraries are files provided by `root`. Maybe there we would not need to specificy the list of files?
For `openjdk` though, there are also external shared libraries, so there a list might be useful.

Or we can just go for a list of files everywhere to be consistent. There won't be that many formuale needing this anyway (maybe 3-4).

**Comment by MikeMcQuaid** [comment]
> I think we should go with linux in the method's name. We have done the same for on_macos/on_linux, where we have specific methods by OS, and not one generic method with a parameter.

Or: it always has to be inside a `on_linux` block? My only concern is if we do at some point decide we need this on macOS that renaming this method is going to be a pain (and also we have no methods except `on_linux` that have `linux` in their name and only `uses_from_macos` for macOS which wouldn't ever be used on Linux anyway).



> Or we can just go for a list of files everywhere to be consistent. There won't be that many formuale needing this anyway (maybe 3-4).

A middle-ground would be a glob. My only concern is this list is pretty long.



> By the way, I think the `root` libraries are files provided by `root`. Maybe there we would not need to specificy the list of files?

If these are all files that are in `lib`: maybe just setting it without a list is enough?

**Comment by sjackman** [comment]
> Or: it always has to be inside a `on_linux` block?

👍

Bike shedding:
```ruby
on_linux do
  ignore_missing_libraries %w[libASImage.so libCling.so libCore.so]
end
```
or perhaps `skip_missing_libraries`.

**Comment by MikeMcQuaid** [comment]
> ```ruby
> on_linux do
>   ignore_missing_libraries %w[libASImage.so libCling.so libCore.so]
> end
> ```

Works for me 👍.

**Comment by maxim-belkin** [comment]
Updated. Formulae can now specify `ignore_missing_libraries` as an array of strings and/or regex-es:
```ruby
ignore_missing_libraries ["libASImage.so", "libCling.so", "libCore.so", %r{libxcb\.so\.\d}]
```

`brew style` is not unhappy with the number of lines in `formula.rb`:
```
$ brew style formula.rb
== formula.rb ==
C: 52:  1: Class has too many lines. [1403/1400]

1 file inspected, 1 offense detected
```

File: Library/Homebrew/linkage_checker.rb [hunk 1]
```diff
@@ -247,7 +264,7 @@ def harmless_broken_link?(dylib)
   # Display a list of things.
   # Things may either be an array, or a hash of (label -> array)
   def display_items(label, things, puts_output: true)
-    return if things.empty?
+    return unless things.present?
```
Comment (sjackman) [inline Library/Homebrew/linkage_checker.rb:267]: I feel like I have slight preference for the former. Why change it?

Comment (maxim-belkin) [reply Library/Homebrew/linkage_checker.rb:267]: I was getting an error here when `broken_dylibs_with_expectations` was `nil`. Let me check if this is still the case.

Comment (maxim-belkin) [reply Library/Homebrew/linkage_checker.rb:267]: Looks like it's no longer the case. I'll revert this change.

**Comment by maxim-belkin** [comment]
Shall we add a special case `:self` to allow missing linkages to libraries provided by the formula itself?

**Comment by sjackman** [comment]
> Shall we add a special case :self to allow missing linkages to libraries provided by the formula itself?

I'm inclined not to add the additional complexity. `ignore_missing_libraries` should be used relatively rarely.

**Comment by jonchang** [comment]
I think this needs a rebase and a quick `brew style --fix`.

**Comment by jonchang** [comment]
:shipit:

**Comment by jonchang** [comment]
Thanks @maxim-belkin! Note that this shouldn't be used in formulae until we tag a release that contains this commit, but it can be tested on e.g. CI builds.

File: Library/Homebrew/linkage_checker.rb [hunk 2]
```diff
@@ -55,18 +55,35 @@ def display_reverse_output
...
+  end
+
+  def unexpected_broken_libs
+    @broken_dylibs.reject { |lib| @formula.allowed_missing_lib? lib }
+  end
+
+  def broken_dylibs_with_expectations
+    output = {}
+    @broken_dylibs.each do |lib|
+      output[lib] = (unexpected_broken_libs.include? lib) ? ["unexpected"] : ["expected"]
```
Comment (MikeMcQuaid) [inline Library/Homebrew/linkage_checker.rb:84]: Could this be an `if`? Bit hard to follow as-is.

File: Library/Homebrew/extend/os/linux/formula.rb [hunk 3]
```diff
@@ -7,11 +7,38 @@ def shared_library(name, version = nil)
...
+    # x -    Name of or a pattern for a library, linkage to which is allowed to be missing.
+    #        Ex. 1: "libONE.so.1"
+    #        Ex. 2: %r{(libONE|libTWO)\.so}
+    self.class.allowed_missing_libraries.any? do |x|
+      case x
+      when Regexp
+        x.match? lib
+      when String
+        lib.include? x
+      end
```
[the remark this answers is not in the document]
Comment (maxim-belkin) [reply Library/Homebrew/extend/os/linux/formula.rb:23]: We can add some logic here but... shall we wait and see if there is a need for that? Right now I suspect there will be 0 use cases for that error-catching "else".

[the remark this answers is not in the document]
Comment (MikeMcQuaid) [reply Library/Homebrew/extend/os/linux/formula.rb:23]: This will fail non-gracefully if there's a need for that. An `else` with a `raise` is minimal code to be maintained but will produce a nicer error message

[the remark this answers is not in the document]
Comment (maxim-belkin) [reply Library/Homebrew/extend/os/linux/formula.rb:23]: > This will fail non-gracefully if there's a need for that.

How would this fail? I thought it'll just quietly ignore it. I'm probably missing something...  Anyways, I agree with adding an `else` but shall it `odie` or `owarn`?

[the remark this answers is not in the document]
Comment (MikeMcQuaid) [reply Library/Homebrew/extend/os/linux/formula.rb:23]: Well, I guess "fail" is the wrong word here but "silently ignore the invalid non-String/non-Regex input" is probably a better one.
